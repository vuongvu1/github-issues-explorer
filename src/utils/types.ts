export type IssueType = Array<{
  id: string;
  title: string;
  number: number;
  createdAt: string;
  author: {
    login: string;
  };
  comments: {
    totalCount: number;
  };
}>;
