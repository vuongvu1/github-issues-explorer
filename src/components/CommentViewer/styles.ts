import styled, { css } from "styled-components";

const SC = {
  Container: styled.nav(
    ({ theme }) => css`
      padding: ${theme.spacing.sm};
      display: flex;
    `
  ),
  Avatar: styled.img(
    ({ theme }) => css`
      width: 40px;
      height: 40px;
      border-radius: 50%;
      margin-right: ${theme.spacing.md};
    `
  ),
  CommentWrapper: styled.div(
    ({ theme }) => css`
      flex-grow: 1;
      border: 1px solid ${theme.palette.dark};
      border-radius: ${theme.borderRadius};
    `
  ),
  CommentHeader: styled.div(
    ({ theme }) => css`
      background-color: ${theme.palette.neutral};
      padding: ${theme.spacing.sm} ${theme.spacing.md};
    `
  ),
  CommentBody: styled.div(
    ({ theme }) => css`
      padding: ${theme.spacing.md};

      p {
        margin: 0;
        margin-bottom: ${theme.spacing.sm};
      }

      img {
        max-width: 100%;
      }

      a {
        text-decoration: none;
      }
    `
  ),
};

export default SC;
