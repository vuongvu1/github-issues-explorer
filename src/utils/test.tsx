import { ReactNode } from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { BrowserRouter } from "react-router-dom";
import { theme } from "src/globalStyles";
import {
  IssueState,
  IssueOrderField,
  OrderDirection,
} from "src/generated/graphql";

const mockStore = configureMockStore();
const store = mockStore({
  repoSlice: {
    url: "https://github.com/reactjs/reactjs.org",
    currentPage: 1,
    searchParams: {
      name: "reactjs.org",
      owner: "reactjs",
      status: IssueState.Open,
      orderBy: IssueOrderField.CreatedAt,
      orderDir: OrderDirection.Desc,
      before: undefined,
      after: undefined,
      first: 10,
      last: undefined,
    },
  },
});

export const customRender = (Component: ReactNode) =>
  render(
    <ThemeProvider theme={theme}>
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Provider store={store}>{Component}</Provider>
        </BrowserRouter>
      </MockedProvider>
    </ThemeProvider>
  );
