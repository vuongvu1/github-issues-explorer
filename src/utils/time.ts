const ONE_YEAR_IN_SECONDS = 31536000;
const ONE_MONTH_IN_SECONDS = 2592000;
const ONE_DAY_IN_SECONDS = 86400;
const ONE_HOUR_IN_SECONDS = 3600;
const ONE_MINUTE_IN_SECONDS = 60;

export const timeSince = (date: Date) => {
  const seconds = Math.floor((new Date().getTime() - date.getTime()) / 1000);

  let interval = seconds / ONE_YEAR_IN_SECONDS;

  if (interval > 1) {
    return `${Math.floor(interval)} years`;
  }

  interval = seconds / ONE_MONTH_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} months`;
  }

  interval = seconds / ONE_DAY_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} days`;
  }

  interval = seconds / ONE_HOUR_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} hours`;
  }

  interval = seconds / ONE_MINUTE_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} minutes`;
  }

  return `${Math.floor(seconds)} seconds`;
};
