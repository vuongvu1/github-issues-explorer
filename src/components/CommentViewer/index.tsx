import { FC } from "react";
import { Converter } from "showdown";
import { AuthorType } from "src/reducers/types";
import { timeSince } from "src/utils/time";
import { Trash as TrashIcon } from "src/assets/icons";
import SC from "./styles";

type Props = {
  id?: string;
  body: string;
  createdAt: string;
  author: AuthorType;
  handleDeleteComment?: (id: string) => void;
};

const converter = new Converter({
  tables: true,
  simplifiedAutoLink: true,
  strikethrough: true,
  tasklists: true,
});

const CommentViewer: FC<Props> = ({
  id,
  body,
  createdAt,
  author,
  handleDeleteComment,
}) => {
  const confirmDelete = () => {
    // eslint-disable-next-line no-restricted-globals
    const isConfirmed = confirm("Are you sure you want to delete this?");

    if (isConfirmed && handleDeleteComment && id) handleDeleteComment(id);
  };

  return (
    <SC.Container>
      <SC.Avatar src={author.avatarUrl} />
      <SC.CommentWrapper>
        <SC.CommentHeader>
          <span>
            <strong>{author?.login}</strong> commented{" "}
            {timeSince(new Date(createdAt))} ago
          </span>
          {handleDeleteComment && (
            <TrashIcon data-testid="delete-btn" onClick={confirmDelete} />
          )}
        </SC.CommentHeader>
        <SC.CommentBody
          dangerouslySetInnerHTML={{
            __html:
              converter.makeHtml(body) || "<i>No description provided.</i>",
          }}
        />
      </SC.CommentWrapper>
    </SC.Container>
  );
};

export default CommentViewer;
