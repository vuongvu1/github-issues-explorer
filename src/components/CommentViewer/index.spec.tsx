import { cleanup, fireEvent } from "@testing-library/react";
import { customRender } from "src/utils/test";
import CommentViewer from "./index";

describe("CommentViewer", () => {
  afterEach(cleanup);

  test("should match snapshot", () => {
    const { container } = customRender(
      <CommentViewer
        author={{
          avatarUrl: "avatarUrl",
          login: "vuongvu",
        }}
        body="Comment body"
        createdAt="2021-04-28T09:02:57Z"
        handleDeleteComment={() => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("should handle delete comment", () => {
    window.confirm = jest.fn(() => true);
    const mockDeleteComment = jest.fn();

    const { getByTestId } = customRender(
      <CommentViewer
        id="id"
        author={{
          avatarUrl: "avatarUrl",
          login: "vuongvu",
        }}
        body="Comment body"
        createdAt="2021-04-28T09:02:57Z"
        handleDeleteComment={mockDeleteComment}
        canDelete
      />
    );

    fireEvent.click(getByTestId("delete-btn"));

    expect(mockDeleteComment).toHaveBeenCalledTimes(1);
  });
});
