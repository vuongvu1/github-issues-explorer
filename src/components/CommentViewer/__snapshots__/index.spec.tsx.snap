// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentViewer should match snapshot 1`] = `
<div>
  <div
    class="sc-gsDJrp fViBsq"
  >
    <img
      class="sc-dkPtyc hVtqgJ"
      src="avatarUrl"
    />
    <div
      class="sc-hKwCoD ijNMMW"
    >
      <div
        class="sc-eCImvq ecNMpG"
      >
        <span>
          <strong>
            vuongvu
          </strong>
           commented
           
          2 days
           ago
        </span>
      </div>
      <div
        class="sc-jRQAMF gfPjGD"
      >
        <p>
          Comment body
        </p>
      </div>
    </div>
  </div>
</div>
`;
